storage/remote/storage.go

```diff
@@ -179,6 +179,11 @@ func (s *Storage) Appender(ctx context.Context) storage.Appender {
 	return s.rws.Appender(ctx)
 }
 
+// LowestSentTimestamp returns the lowest sent timestamp across all queues.
+func (s *Storage) LowestSentTimestamp() int64 {
```

Comment:
Why do you need that?

Prometheus does not receive samples in order. We could have a sample from a new metric which is one hour back from the already-sent timestamps.

Reply:
I describe this a bit in the PR description:

> Truncation of the WAL cooperates on a best-effort case with Remote Write. Every time the WAL is truncated, the minimum timestamp of data to truncate is determined by the lowest sent timestamp of all samples across all remote_write endpoints. This gives loose guarantees that data from the WAL will not try to be removed until the maximum sample lifetime passes or remote_write starts functioning.

If we don't use this, we'll need to find another way to determine what the minimum timestamp for truncation could be. It could just be the biggest timestamp from the least append, which I _think_ is what the TSDB uses.

FWIW, though, this is finding the smallest value for the latest sent timestamp (largest) across all queue_managers. It can help prevent deleting data when one of the remote_writes is lagging. Appending a sample hours back shouldn't affect the logic.

Reply:
I realize I've mis-explained this in the PR description. Trying to communicate "lowest value among the highest timestamps" is confusing, which is probably a good reason to change this to something that's easier to think about.

Reply:
@roidelapluie unfortunately the right way to do this has to be implemented still, see item 3 here: https://github.com/prometheus/prometheus/issues/6333#issue-524638121

That would allow both sending of samples that were missed if prometheus shut down and had fallen behind, as well as proper determination of what segments are up for checkpointing/truncation for the agent.